Fix NNX jit static args with in_shardings issue #4989

mohsinm-dev opened the pull request:
## Problem

When using `nnx.jit` with static arguments and `in_shardings`, users encountered errors:

1. Not providing shardings for static args → "Tuple arity mismatch" error
2. Providing shardings for static args → "pjit in_shardings specification must be a tree prefix" error

## Solution

Fixed the sharding handling logic to properly filter out shardings for static argument positions before passing them to JAX's jit function.

## Changes

- Modified `JitWrapped.__init__` in `compilation.py` to drop shardings for static positions
- Added comprehensive tests to verify both scenarios work correctly

## Testing

- Added `tests/nnx/jit_static_sharding_test.py` with tests for both failing scenarios
- Verified no regressions in existing functionality

Fixes #4989

samanklesaria commented on tests/nnx/transforms_test.py:
```diff
@@ -420,6 +420,58 @@ def f(m: nnx.Linear, x):
+    # Verify static arg toggles result
+    self.assertNotEqual(y_relu, y_no_relu)
+
+  def test_jit_static_with_extra_static_entry(self):
```
I don't think that passing a sharding spec for a static argument is something that  `jax.jit` supports. When I try compilation here with `jax.jit` instead of `nnx.jit`,  I get `got a tuple or list of length 3 for an args tuple of length 2`. So it seems like something `nnx.jit` shouldn't support either.

samanklesaria commented on flax/nnx/transforms/compilation.py:
```diff
@@ -362,12 +408,34 @@ def __init__(
+      )
+
+    def _drop_static_from_sequence(seq):
+      # Keep only shardings for dynamic (non-static) positions.
```
Seems like this would be easier with a list comprehension one liner: 
`type(seq)([v for i,v in in enumerate(seq) if i not in self._static_positions])`.

samanklesaria commented on flax/nnx/transforms/compilation.py:
```diff
@@ -405,14 +471,55 @@ def __get__(self, obj, objtype=None):
+      # Determine number of positional args and number of dynamic args.
+      num_args = len(args)
+      static_set = set(self._static_positions)
+      num_dynamic = num_args - sum(1 for i in range(num_args) if i in static_set)
```
Is this not just `num_args - len(static_set)`?

samanklesaria commented on flax/nnx/transforms/compilation.py:
```diff
@@ -405,14 +471,55 @@ def __get__(self, obj, objtype=None):
+      # None for static positions.
+      full_prefix: list[tp.Any] = []
+      dyn_idx = 0
+      for i in range(num_args):
```
It might be slightly cleaner to use an iterator here. 
```python
shard_iter = iter(dyn_specs)
tuple(None if i in static_set else next(dyn_specs) for i in range(num_args))
```

samanklesaria commented on tests/nnx/transforms_test.py:
```diff
@@ -420,6 +420,58 @@ def f(m: nnx.Linear, x):
+    scale = 0.5
+
+    x_sharding = jax.sharding.NamedSharding(mesh, jax.sharding.PartitionSpec('data'))
+    f = nnx.jit(fn, in_shardings=(x_sharding, None), static_argnums=(2,))
```
We should test when `in_shardings` needs to be broadcast as well. Add another test case for this.

mohsinm-dev commented on flax/nnx/transforms/compilation.py:
```diff
@@ -405,14 +471,55 @@ def __get__(self, obj, objtype=None):
+        # past available entries.
+      else:
+        # Single spec: broadcast to all dynamic args
+        dyn_specs = [self.in_shardings] * max(num_dynamic, 0)
```
The explicit broadcasting is not needed here.

mohsinm-dev replied:
@vfdev-5 Maybe I am incorrect, but the coupling exists because sharding information drives the tree transformation decision, not just the final JAX call.

`extract.to_tree` uses the prefix to decide which arguments to transform:
- Non-None prefix values → transform argument to NodeStates
- None prefix values → leave argument unchanged (static)

1. `broadcast_prefix(single_sharding, args)` → ['SHARD', 'SHARD', 'SHARD'] (applies to ALL positions, including static - incorrect)
2. `broadcast_prefix((shard1, shard2, None), args)` → ['SHARD1', 'SHARD2', None] (correctly marks static position)

And I don't think so without sharding information in the prefix, `extract.to_tree` has no way to distinguish which arguments should be transformed from those that should be preserved as static. The transformation step is where the sharding information is consumed.

mohsinm-dev commented:
@vfdev-5 Thanks for the suggestion. The current approach works as well, but, as you suggested, depending on the JAX, it is better, and we stay consistent with JAX's behavior for edge cases, such as mixed static_argnums and static_argnames.
I've implemented your approach and added comprehensive tests, all of which passed successfully.

vfdev-5 commented on flax/nnx/transforms/compilation.py:
```diff
@@ -128,16 +130,61 @@ def __call__(self, *pure_args, **pure_kwargs):
     pure_args_out, pure_kwargs_out, pure_out = extract.to_tree(
       (args_out, kwargs_out, out),
-      prefix=(self.in_shardings, self.kwarg_shardings, self.out_shardings),
+      prefix=(None, self.kwarg_shardings, self.out_shardings),
```
Let's revert this as this can be a BC-breaking.

mohsinm-dev commented on flax/nnx/transforms/compilation.py:
```diff
@@ -405,14 +479,31 @@ def __get__(self, obj, objtype=None):
     return functools.partial(self, obj)
 
   def _get_pure_args_kwargs(self, args, kwargs):
+    # Build prefix for extract.to_tree
```
But we need to build the prefix with None at static positions. I think it is essential for the fix.
